tests/test_grpc_functions.py

```diff
@@ -0,0 +1,248 @@
+__copyright__ = '''
+Copyright 2018 the original author or authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+'''
+__author__ = 'David Turanski'
+
+import grpc
+import unittest
+import subprocess
+import os
+import uuid
+import time
+
+from invoker import function_pb2_grpc as function
+from invoker import function_pb2 as message
+
+# TODO: Make this portable
+PYTHON2 = "~/miniconda2/bin/python"
+
+class GrpcFunctionTest(unittest.TestCase):
+    """
+    Assumes os.getcwd() is the project base directory
+    """
+    @classmethod
+    def setUpClass(cls):
+        # TODO: Make this portable
+        cls.workingdir = os.path.abspath("./invoker")
+        cls.command = "%s function_invoker.py" % PYTHON2
+
+    def setUp(self):
+        pass
+
+    def tearDown(self):
+        print("killing %s" % self.process.pid)
+        self.process.kill()
+
+    def test_upper(self):
+        port = find_free_port()
+        env = {
+            'PYTHONPATH': '%s/tests/functions:$PYTHONPATH' % os.getcwd(),
+            'GRPC_PORT': str(port),
+            'FUNCTION_URI': 'file://%s/tests/functions/upper.py?handler=handle' % os.getcwd()
+        }
+
+        self.process = subprocess.Popen(self.command,
+                                        cwd=self.workingdir,
+                                        shell=True,
+                                        env=env,
+                                        preexec_fn=os.setsid,
+                                        )
+        time.sleep(0.5)
```

Comment:
This looks like a race. On a slow (CI) system, 0.5 won't be enough and the test could presumably fail. Is there a better way?

Reply:
Brittle but how to tell the server is ready for requests? We could add a ping() interface to the gRPC definition.